### PR TITLE
Do MIOpen validation tests and MLIR kernels tuning weekly

### PR DESCRIPTION
**Pull request by yiqian1**
- Add a "weekly" parameter to Jenkinsfile. The combination of "staticLib" and "weekly" does:

  Validate MIOpen Resnet50 with tuning and archive the user perfDb.
  Run MIopen tests without tuning.

- Reduce the matrix size of the "Test MIOpen config" to 2x2

- Add a "--tune-all" command line option to miopen_validate.sh which does tuning on all Resnet50 configs.

- Set Xdlops based on GPU type (via rocm_agent_enumerator), thus we can separate MIOpen tests on private CI and public CI.

A successful weekly CI is:
http://ml-ci.amd.com:21096/blue/organizations/jenkins/MLIR%2Ftest_weeklyCI/detail/test_weeklyCI/9/pipeline/276/

### CODE REVIEW

File: mlir/utils/jenkins/Jenkinsfile [hunk 1]
```diff
@@ -236,13 +241,10 @@ pipeline {
...
-                                equals expected: true, actual: params.nightly;
-                                anyOf {
-                                    equals expected: true, actual: params.perfTest;
-                                    equals expected: true, actual: params.canXdlops;
-                                } }
-                            }
+                            when { anyOf {
+                                    equals expected: true, actual: params.nightly;
+                                    equals expected: true, actual: params.weekly;
+                            }}
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile:250]: Let's keep the control constants in here

File: mlir/utils/jenkins/Jenkinsfile [hunk 2]
```diff
@@ -267,12 +269,10 @@ pipeline {
...
-                            when { allOf {
-                                equals expected: true, actual: params.nightly;
-                                anyOf {
-                                equals expected: true, actual: params.perfTest;
-                                equals expected: true, actual: params.canXdlops;
-                            } } }
+                            when { anyOf {
+                                    equals expected: true, actual: params.nightly;
+                                    equals expected: true, actual: params.weekly;
+                            }}
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile:281]: Same deal - let's make sure we can run individual parts of nightly

File: mlir/utils/jenkins/Jenkinsfile [hunk 3]
```diff
@@ -281,6 +281,29 @@ pipeline {
...
+                                sh 'ldconfig'
+                                dir('MIOpen/build/') {
+                                    sh 'date --utc +%Y-%m-%d > tuning-date'
+                                    sh """
+                                    bash ${WORKSPACE}/mlir/utils/jenkins/miopen-tests/miopen_validate.sh --tune-all\
+                                       < ${WORKSPACE}/mlir/utils/jenkins/miopen-tests/resnet50-miopen-configs"""
+                                    sh 'ls -l MIOpenUserDB'
+                                }
+                                // Save user database  for nightly jobs
+                                archiveArtifacts artifacts: 'MIOpen/build/MIOpenUserDB/**, MIOpen/build/tuning-date',\
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile:309]: 1) This only runs weekly, right?

2) Have you tested to make sure this won't conflict with performance report archiving?

File: mlir/utils/jenkins/Jenkinsfile [hunk 4]
```diff
@@ -354,27 +377,35 @@ pipeline {
...
                 stage("Copy MIOpen with MLIR kernels") {
                     steps {
                         sh 'rm -rf MIOpen'
                         unstash name: "MIOpen-MLIR-kernels"
+                        //copyArtifacts filter: 'MIOpen/build/MIOpenUserDB,MIOpen/build/tuning-date',\
+                        //                       optional: true,\
+                        //                       flatten: true,\
+                        //                       projectName: "/${JOB_NAME}",\
+                        //                       selector: lastSuccessful(),\
+                        //                       target: 'MIOpen/build/'
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile:413]: Why's this commented out?

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh [hunk 5]
```diff
@@ -78,8 +94,8 @@ function construct_fixed_args() {
        exit 2
     fi
     case "$DTYPE" in
-        fp16 ) FIXED_CONFIG_ARGS+=("convfp16") ;;
-        fp32 ) FIXED_CONFIG_ARGS+=("conv") ;;
+        fp16 ) FIXED_CONFIG_ARGS=("convfp16") ;;
```
Comment (krzysz00) [inline mlir/utils/jenkins/miopen-tests/miopen_validate.sh:111]: Why this change?

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh [hunk 6]
```diff
@@ -125,8 +136,12 @@ function setup_environment() {
         4) MIOPEN_DEBUG_FIND_ONLY_SOLVER=ConvMlirIgemmWrW ;;
         *) echo "$0: Unsupported direction flag $DIRECTION"; exit 2
     esac
-    if [[ $XDLOPS == 1 ]]; then
+    local xdlops
+    xdlops=$(/opt/rocm/bin/rocm_agent_enumerator |grep gfx908)
```
Comment (krzysz00) [inline mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: Is this really the right approach? It feels like we'd want to have xdlops be something we can pass in instead of relying on autodetection.

That is, I'm curious about the logic here.

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh [hunk 7]
```diff
@@ -138,8 +153,6 @@ function run_tests() {
         exit 3
     fi
 
-    clean_miopen_caches
```
Comment (krzysz00) [inline mlir/utils/jenkins/miopen-tests/miopen_validate.sh:141]: ?

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh [hunk 8]
```diff
@@ -160,12 +173,35 @@ function run_tests() {
...
+            done
+        done
+    done
+}
+
+function main() {
+    clean_miopen_caches
+    parse_options "$@"
+    get_configs
+    if [[ $TUNEALL == 1 ]]; then
```
Comment (krzysz00) [inline mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: The tuneall machinery is so you can have the Jenkinsfile run all the tuning on one node?

**Review by krzysz00 — COMMENTED** [review]
I'm not entirely understanding how tuning will interact with the rest of nightly CI - ex. with the performance reports, especially since the important bit seems to be commented out. Can you clarify?

File: mlir/utils/jenkins/Jenkinsfile (hunk 2 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:281]: > I'm not entirely understanding how tuning will interact with the rest of nightly CI - ex. with the performance reports, especially since the important bit seems to be commented out. Can you clarify?

The tuning results will be used by the performance reports in nightly CI. This will be in another PR.

File: mlir/utils/jenkins/Jenkinsfile (hunk 3 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:309]: Yes. This runs weekly. I haven't tested this with performance reports yet, but I don't think they will conflict, as they are different CI jobs.

File: mlir/utils/jenkins/Jenkinsfile (hunk 4 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:413]: This will be in another PR.

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh (hunk 5 above)
Comment (yiqian1) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh:111]: To reset FIXED_CONFIG_ARGS every time construct_fixed_args() is called. For the tune-all case, construct_fixed_args() is called multiple times.

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh (hunk 6 above)
Comment (yiqian1) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: The autodetection allows us to run xdlops tests on private CI and non-xdlops on public CI. That's what the performance test is doing. Thus we can test on more CI nodes and have a more balanced work on private and public CI.

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh (hunk 7 above)
Comment (yiqian1) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh:141]: Moved clean_miopen_caches to main().

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh (hunk 8 above)
Comment (yiqian1) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: Yes. I want the tuning to generate one user perfDb which I can use for performance report. If the tuning is run in parallel, there will be multiple perfDbs at different nodes, making them hard to be used.

File: mlir/utils/jenkins/Jenkinsfile (hunk 1 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:250]: What do you mean?
With this change, the stage can be run for nightly+staticLib (MIOpen tests), nightly+perfTest, or weekly+staticLib. If canXdlops=true, it is on a private CI node; otherwise, on a public CI node.

File: mlir/utils/jenkins/Jenkinsfile (hunk 2 above)
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile:281]: Re the control constants, what I mean is we still want the `perfTest` variable so that we can run just performance tests, or something analogous to it.

File: mlir/utils/jenkins/Jenkinsfile (hunk 3 above)
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile:309]: Yeah, that's true, but currently `weekly` duplicates a lot of `nightly` and so we'll be getting things like weekly performance regressions.

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh (hunk 5 above)
Comment (krzysz00) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh:111]: Makes sense

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh (hunk 6 above)
Comment (krzysz00) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: Good point, and this makes sense as a change.

Comment (krzysz00) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: Though it might be nice to retain the ability to override `xdlops` for manual testing

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh [hunk 9]
```diff
@@ -59,13 +63,25 @@ function parse_options() {
     done
 
     # Check required args
-    [[ $got_layout == 1 && $got_direction == 1 && $got_dtype == 1 ]] || usage
+    [[ $got_layout == 1 && $got_direction == 1 && $got_dtype == 1 ]] || [[ $TUNEALL == 1 ]] || usage
```
Comment (krzysz00) [inline mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: I think you can have only one call to `[[`, but that's a matter of taste

File: mlir/utils/jenkins/Jenkinsfile (hunk 1 above)
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile:250]: I mean that we don't have this stage restricted to (perfTest || staticLib) && (nightly || weekly)

**Review by krzysz00 — COMMENTED** [review]
I've got a few minor nitpicks but otherwise I'm happy with this

File: mlir/utils/jenkins/Jenkinsfile (hunk 1 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:250]: Changed it to weekly && staticLib.

**Review by jerryyin — DISMISSED** [review]
Based on the existing [run](http://ml-ci.amd.com:21096/blue/organizations/jenkins/MLIR%2Ftest_weeklyCI/detail/test_weeklyCI/9/pipeline/276), my high level feedbacks are:

- In weekly CI, MIOpen resnet 50 configs shall test with both tuning on and off, yet you only have tuning off
- In nightly CI, we need at least 2 configs of all alternatives to make sure its minimal test coverage

**Comment by yiqian1** [comment]
> Based on the existing [run](http://ml-ci.amd.com:21096/blue/organizations/jenkins/MLIR%2Ftest_weeklyCI/detail/test_weeklyCI/9/pipeline/276), my high level feedbacks are:
> 
> * In weekly CI, MIOpen resnet 50 configs shall test with both tuning on and off, yet you only have tuning off

Tuning is tested in the "Test MIOpen config with tuning" stage.

> * In nightly CI, we need at least 2 configs of all alternatives to make sure its minimal test coverage

This will be in another PR.

**Comment by jerryyin** [comment]
> Tuning is tested in the "Test MIOpen config with tuning" stage.

That stage alone is costing more than 3 hours, is that the best way to do it?

Tuning on private CI will take even longer time. That's why we don't want nightly CI to do it. I'm not too worried about the time as it is run on a weekly base. This stage includes tuning for NCHW and NHWC. I can make them in parallel. That's probably the only thing I can do.

> This will be in another PR.

May I ask what the ETA for the other PR: In the mean time we'd lose coverage if anything break MIOpen integration (and will only know until weekly tests). Therefore desirably the next PR needs to be out as soon as possible.

I want to make it out this week. The only thing that is not covered by the current nightly CI is tuning for NHWC. The performance test most likely will fail if any tuning for NCHW fails.

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh (hunk 6 above)
Comment (yiqian1) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: Good idea. I think I will have it in the next PR.

Comment (krzysz00) [reply mlir/utils/jenkins/miopen-tests/miopen_validate.sh]: Could you add it here?

The way I'd do this is to have the xdlops autodetection in options parsing - if the user doesn't pass in `--xdlops` or `--no-xdlops`, call the agent enumerator and detect a setting.

File: mlir/utils/jenkins/miopen-tests/miopen_validate.sh [hunk 10]
```diff
@@ -59,13 +66,36 @@ function parse_options() {
...
+        case "$DTYPE" in
+            fp16 | fp32 ) ;;
+            * ) echo "$0: Invalid dtype $DTYPE"; usage ;;
+        esac
+    fi
+
+    # Detect XDLOPS if not specified
+    local xdlops
+    if [[ $XDLOPS == -1 ]]; then
+      xdlops=$(/opt/rocm/bin/rocm_agent_enumerator |grep gfx908)
```
Comment (krzysz00) [inline mlir/utils/jenkins/miopen-tests/miopen_validate.sh:82]: My brain is saying `grep -q` and checking the exit code in the if statement

That is dropping `xdlops` and going
```
if /opt/rocm/bin/rocm_agent_enumerator | grep -q gfx908; then
...
```

but that's a really tiny nitpick

File: mlir/utils/jenkins/Jenkinsfile (hunk 1 above)
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile:250]: Ok, just to make sure, this test needs to run in every nightly, but only in weeklies where we're doing static library tests?

`nightly || (weekly && staticLib)` doesn't feel right.

And, as a followup, why is this now `staticLib` and not `perfTest`?

**Review by krzysz00 — COMMENTED** [review]
One nitpick, one actual question, otherwise looks fine

File: mlir/utils/jenkins/Jenkinsfile (hunk 1 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:250]: > Ok, just to make sure, this test needs to run in every nightly, but only in weeklies where we're doing static library tests?
> 
This is true for now. Eventually this test is only run in weeklies. Nightlies will only run two selected MIOpen configs.

> `nightly || (weekly && staticLib)` doesn't feel right.
> 

When I'm done with the new nightly CI, this will become (weekly && staticLib)

> And, as a followup, why is this now `staticLib` and not `perfTest`?

Because it's not part of performance tests.

Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile:250]: Ah. In which case, so as to not break the selective CI setup, can you make this `(nightly || weekly) && staticLib`

Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:250]: Sorry. I said wrong. This test is run (nightly&&perfTest) || (nightly&&staticLib) || (weekly&&staticLib). The condition (perfTest || staticLIb) has been specified for the top-level stage "Static Library". Since perfTest is only done in nightlies, I disallow it for weeklies.

Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile:250]: Ah. That makes sense, thanks for clarifying!

**Review by krzysz00 — APPROVED** [review]
LGTM, overall
